neuralmonkey/decoders/decoder.py

```diff
@@ -308,88 +337,91 @@ def _logit_function(self, rnn_output):
 
-            for step in range(1, len(inputs)):
-                tf.get_variable_scope().reuse_variables()
+        Arguments:
+            inputs: The decoder inputs. If runtime_mode=True, only the first
+                    input is used.
+            scope: Scope for logit function and output projection variables
+            runtime_mode: Boolean flag whether the decoder is running in
+                          runtime mode (with loop function).
+        """
+        ## CHANGE:
+        ## EMBEDDED INPUTS NOW OF SHAPE TIME x BATCH x RNN_SIZE
+        ## EMBEDDED GO_SYMBOLS ARE 1 x BATCH x RNN_SIZE
 
-                if runtime_mode:
-                    current_input = self._loop_function(output)
-                else:
-                    current_input = inputs[step]
+        assert runtime_mode == scope.reuse
+        logit, output, state = self.decoding_step(
+            inputs[0], self.initial_state, scope)
 
-                ## N-th decoding step
-                contexts = [a.attention(state) for a in att_objects]
-                output = self.output_projection(current_input, state, contexts)
-                _, state = cell(tf.concat(1, [current_input] + contexts), state)
+        output_logits = [logit]
+        rnn_outputs = [output]
+        rnn_states = [self.initial_state, state]
 
-                logit = self._logit_function(output)
+        assert runtime_mode == scope.reuse
+        scope.reuse_variables()
+        assert scope.reuse
 
-                output_logits.append(logit)
-                rnn_outputs.append(output)
-                rnn_states.append(state)
+        for step in range(1, self.max_output_len + 1):
 
             if runtime_mode:
-                for i, a in enumerate(att_objects):
-                    attentions = a.attentions_in_time[-len(inputs):]
-                    alignments = tf.expand_dims(tf.transpose(
-                        tf.pack(attentions), perm=[1, 2, 0]), -1)
+                # NOTE loop function must never leave this scope
+                # because the _logit_function is scope-sensitive,
+                # meaning it would create a new set of parameters
+                # in a different scope
+                current_input = self._loop_function(output, scope)
+            else:
+                current_input = inputs[step]
+
+            logit, output, state = self.decoding_step(current_input, state,
+                                                      scope)
 
-                    tf.image_summary("attention_{}".format(i), alignments,
-                                     collections=["summary_val_plots"],
-                                     max_images=256)
+            output_logits.append(logit)
+            rnn_outputs.append(output)
+            rnn_states.append(state)
 
         return rnn_outputs, rnn_states, output_logits
 
 
-    def _init_summaries(self):
-        """Initialize the summaries of the decoder
+    def _visualize_alignments(self):
```

Comment:
Já bych tomu neříkal alignments, Ondra se chystá napsat použití předtrénovaného alignmentu, tak bych radši pořádně všude odlišoval, co je alignment a co je attention, aby někdy v budoucnu nedošlo k nějakým zmatkům.